packages/react/src/host/RemoteRenderer.tsx

```diff
@@ -1,41 +1,55 @@
-import {memo} from 'react';
+import {cloneElement, memo, ReactElement, useMemo} from 'react';
 import {KIND_COMPONENT, KIND_TEXT, RemoteReceiver} from '@remote-ui/core';
 
 import type {Controller} from './controller';
 import {useAttached} from './hooks';
-import {RemoteText} from './RemoteText';
-import {RemoteComponent} from './RemoteComponent';
+import {renderText as defaultRenderText} from './RemoteText';
+import {renderComponent as defaultRenderComponent} from './RemoteComponent';
+import {RemoteRendererContext} from './context';
+import type {RemoteComponentProps, RemoteTextProps} from './types';
 
-interface Props {
+export interface RemoteRendererProps {
   receiver: RemoteReceiver;
   controller: Controller;
+  renderComponent?: (props: RemoteComponentProps) => ReactElement;
+  renderText?: (props: RemoteTextProps) => ReactElement;
 }
 
-export const RemoteRenderer = memo(({controller, receiver}: Props) => {
-  const {children} = useAttached(receiver, receiver.attached.root)!;
+export const RemoteRenderer = memo(
+  ({
+    controller,
+    receiver,
+    renderComponent = defaultRenderComponent,
+    renderText = defaultRenderText,
+  }: RemoteRendererProps) => {
+    const {children} = useAttached(receiver, receiver.attached.root)!;
+    const renderContextValue = useMemo(() => ({renderText, renderComponent}), [
+      renderText,
+      renderComponent,
+    ]);
 
-  return (
-    <>
-      {children.map((child) => {
-        switch (child.kind) {
-          case KIND_COMPONENT:
-            return (
-              <RemoteComponent
-                key={child.id}
-                component={child}
-                receiver={receiver}
-                controller={controller}
-                __type__={(controller.get(child.type) as any)?.__type__}
-              />
-            );
-          case KIND_TEXT:
-            return (
-              <RemoteText key={child.id} text={child} receiver={receiver} />
-            );
-          default:
-            return null;
-        }
-      })}
-    </>
-  );
-});
+    return (
+      <RemoteRendererContext.Provider value={renderContextValue}>
+        {children.map((child) => {
+          let element: ReactElement | null;
+          switch (child.kind) {
+            case KIND_COMPONENT:
+              element = renderComponent({
+                component: child,
+                receiver,
+                controller,
+              });
+              break;
+            case KIND_TEXT:
+              element = renderText({text: child, receiver});
+              break;
+            default:
+              element = null;
+              break;
+          }
+          return element ? cloneElement(element, {key: child.id}) : null;
```

Comment:
Same question regarding the need to clone the element

Reply:
It's about the key